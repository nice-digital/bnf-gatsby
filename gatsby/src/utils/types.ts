import { type Primitive, type Merge, type Simplify } from "type-fest";

export type BuiltIns = Primitive | Date | RegExp;

export interface RecordSection extends Slug {
	order: number;
	title: string;
	content: string;
}

export interface MenuPageLink {
	href: string;
	title: string;
}

/** Represents the structure of JSON files used for meta descriptions:
 * a map of URL slug to string descriptions for BNF and BNFC */
export type MetaDescriptionsMap = Record<
	string,
	{ bnf: string | null; bnfc: string | null } | undefined
>;

export interface Slug {
	slug: string;
}

export interface SlugAndTitle extends Slug {
	title: string;
}

export type WithSlug<T extends object> = T & Slug;

/**
 * Add a `slug: string` property to each of the given `SlugTypes` recursively within the given type `Base`.
 *
 * This type is useful for augmenting raw feed types with `slug` properties when they're queried via GraphQl.
 * This is because we add slugs using the `@slug` directive in the GraphQL schema, so there are no slugs on the raw feed types.
 * It avoids having to do nested union types to add a `slug` property in each file that queries GraphQL.
 *
 * This type will often be used in tandem with `QueryResult` in practice to correctly type the results of GraphQL queries _and_ to add slugs where needed.
 *
 * 	@example To add slugs to all `FeedMedicinalForm` types within `FeedMedicinalForms`:
 * 		type MedicinalFormsWithSlugs = WithSlugDeep<FeedMedicinalForms, FeedMedicinalForm>
 * */
export type WithSlugDeep<
	Base extends object,
	SlugTypes extends object
> = Simplify<
	Merge<
		Base extends SlugTypes ? WithSlug<Base> : Base,
		{
			[Key in keyof Base]: Base[Key] extends (infer U)[] | undefined
				? U extends object
					? WithSlugDeep<U, SlugTypes>[] | undefined
					: Base[Key]
				: Base[Key] extends object
				? WithSlugDeep<Base[Key], SlugTypes>
				: Base[Key] extends object | undefined
				? WithSlugDeep<NonNullable<Base[Key]>, SlugTypes> | undefined
				: Base[Key];
		}
	>
>;

/**
 * Wrapper type for types returned from GraphQL queries.
 *
 * It does the following:
 * - replaces `undefined` with `null` (GraphQL queries return `null` for empty properties)
 * - makes arrays required (not `null` or `undefined`)
 * */
export type QueryResult<Base> = Base extends NonNullable<BuiltIns>
	? Base
	: Base extends BuiltIns
	? NonNullable<Base> | null
	: Base extends (infer U)[]
	? NonNullable<QueryResult<U>[]>
	: Base extends object
	? {
			[Key in keyof Base]-?: NonNullable<Base[Key]> extends unknown[]
				? NonNullable<QueryResult<Base[Key]>>
				: QueryResult<Base[Key]>;
	  }
	: unknown;
