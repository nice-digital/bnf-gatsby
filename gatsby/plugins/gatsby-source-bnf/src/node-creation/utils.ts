import type { NodeInput, SourceNodesArgs } from "gatsby";

export type BnfNodeType = `Bnf${string}`;

export type TypedNodeInput<
	TNodeType extends BnfNodeType,
	TNodeContent
> = NodeInput &
	TNodeContent & {
		// Internal Gatsby node stuff
		internal: {
			type: TNodeType;
		} & NodeInput["internal"];
	};

/**
 * Helper method to create a BNF-specific Gatsby GraphQL Node, that is, a node with a type starting with `Bnf`.
 *
 * @param nodeContent The content of the node to create, without any of the Gatsby internals
 * @param type The name of the node type, starting with `Bnf` e.g. `BnfDrug`
 * @param sourceNodesArgs The arguments to `sourceNodes` from Gatsby
 * @template TNodeContent The type of the node content
 */
export const createBnfNode = <TNodeContent extends { id: string }>(
	nodeContent: TNodeContent,
	type: BnfNodeType,
	{ createContentDigest, actions: { createNode } }: SourceNodesArgs
): void => {
	const content = JSON.stringify(nodeContent),
		contentDigest = createContentDigest(content);

	const node: TypedNodeInput<typeof type, TNodeContent> = {
		...nodeContent,
		internal: {
			type,
			content,
			contentDigest,
		},
	};

	createNode(node);
};
