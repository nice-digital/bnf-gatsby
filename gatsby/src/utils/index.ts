export * from "./types";

/**
 * Returns whether the given object is truthy or not but casted to the object's type.
 *
 * This casting allows filtering of types that are potentially falsy:
 *
 * @example
 * 	const list: (SomeType | undefined)[] = []
 * 	someList.filter(isTruthy).map(a: SomeType => {
 *  		// Do something
 *  	})
 * */
export const isTruthy = <TMaybe>(maybeT: TMaybe | null): maybeT is TMaybe =>
	!!maybeT;
