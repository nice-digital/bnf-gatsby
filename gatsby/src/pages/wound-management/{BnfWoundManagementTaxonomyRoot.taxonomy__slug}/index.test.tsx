import { render, screen, within, waitFor } from "@testing-library/react";
import React from "react";

import { decapitalize } from "@/utils";

import WoundManagementTaxonomyPage, {
	type WoundManagementTaxonomyPageProps,
} from "./";

const props: WoundManagementTaxonomyPageProps = {
	data: {
		allBnfWoundManagementTaxonomyRoot: {
			taxonomies: [
				{
					taxonomy: {
						title: "Basic wound contact dressings",
						slug: "basic-wound-contact-dressings",
					},
				},
				{
					taxonomy: {
						title: "Advanced wound dressings",
						slug: "advanced-wound-dressings",
					},
				},
			],
		},
		bnfWoundManagementTaxonomyRoot: {
			taxonomy: {
				title: "Advanced wound dressings",
				slug: "advanced-wound-dressings",
				text: "<p> Advanced wound dressings can be used for both acute and chronic wounds. Categories for dressings in this section start with the least absorptive, moisture-donating hydrogel dressings, followed by increasingly more absorptive dressings. </p>",
				childTaxonomies: [
					{
						title: "Hydrogel dressings",
						slug: "hydrogel-dressings",
						text: "<p> Hydrogel dressings are most commonly supplied as an amorphous, cohesive topical application that can take up the shape of a wound.</p>",
						productGroups: [],
						childTaxonomies: [
							{
								title: "Sodium hyaluronate dressings",
								slug: "sodium-hyaluronate-dressings",
							},
							{
								title: "Hydrogel sheet dressings",
								slug: "hydrogel-sheet-dressings",
							},
							{
								title: "Hydrogel application (amorphous)",
								slug: "hydrogel-application-amorphous",
							},
						],
					},
					{
						title: "Vapour-permeable films and membranes",
						slug: "vapour-permeable-films-and-membranes",
						text: "<p> Vapour-permeable films and membranes allow the passage of water vapour and oxygen but are impermeable to water and micro-organisms, and are suitable for lightly exuding wounds.</p>",
						productGroups: [],
						childTaxonomies: [
							{
								title: "Non-woven fabric dressing with viscose-rayon pad",
								slug: "non-woven-fabric-dressing-with-viscose-rayon-pad",
							},
							{
								title:
									"Vapour–permeable transparent film dressing with adhesive foam border",
								slug: "vapour-permeable-transparent-film-dressing-with-adhesive-foam-border",
							},
							{
								title: "Vapour–permeable transparent, adhesive film dressing.",
								slug: "vapour-permeable-transparent-adhesive-film-dressing",
							},
							{
								title:
									"Vapour-permeable Adhesive Film Dressing (Semi-permeable Adhesive Dressing)",
								slug: "vapour-permeable-adhesive-film-dressing-semi-permeable-adhesive-dressing",
							},
						],
					},
					{
						title: "Capillary-acting dressings",
						slug: "capillary-acting-dressings",
						text: null,
						productGroups: [],
						childTaxonomies: [],
					},
				],
			},
		},
	},
};

const childTaxonomyListName = `List of ${props.data.bnfWoundManagementTaxonomyRoot.taxonomy.title.toLowerCase()}`;

describe("Wound management taxonomy page", () => {
	it("should render the page title with the expected text", async () => {
		render(<WoundManagementTaxonomyPage {...props} />);
		await waitFor(() => {
			expect(document.title).toStartWith(
				props.data.bnfWoundManagementTaxonomyRoot.taxonomy.title
			);
		});
	});

	it("should render meta description", async () => {
		render(<WoundManagementTaxonomyPage {...props} />);
		await waitFor(() => {
			// eslint-disable-next-line testing-library/no-node-access
			expect(document.querySelector("meta[name=description]")).toHaveAttribute(
				"content",
				`This wound management topic describes the options that are currently recommended for ${decapitalize(
					props.data.bnfWoundManagementTaxonomyRoot.taxonomy.title
				)}`
			);
		});
	});

	it("should match page snapshot", () => {
		const { container } = render(<WoundManagementTaxonomyPage {...props} />);

		expect(container).toMatchSnapshot();
	});

	it("should add content start skip link target id to page header", () => {
		render(<WoundManagementTaxonomyPage {...props} />);
		const heading1 = screen.getByRole("heading", {
			level: 1,
		});
		// eslint-disable-next-line testing-library/no-node-access
		expect(heading1.parentElement).toHaveProperty("id", "content-start");
	});

	it("should render the page heading with the expected text", () => {
		render(<WoundManagementTaxonomyPage {...props} />);
		expect(screen.getByRole("heading", { level: 1 })).toHaveTextContent(
			props.data.bnfWoundManagementTaxonomyRoot.taxonomy.title
		);
	});

	it("should render a navigation list of sub sections", () => {
		render(<WoundManagementTaxonomyPage {...props} />);
		expect(
			screen.getByRole("navigation", {
				name: "Wound management product pages",
			})
		).toBeInTheDocument();
	});

	it("should render correct number of sections in the sub nav", () => {
		render(<WoundManagementTaxonomyPage {...props} />);
		const subNav = screen.getByRole("navigation", {
			name: "Wound management product pages",
		});
		expect(within(subNav).getAllByRole("listitem")).toHaveLength(
			props.data.allBnfWoundManagementTaxonomyRoot.taxonomies.length
		);
	});

	it("should render a list of child taxonomies", () => {
		render(<WoundManagementTaxonomyPage {...props} />);
		expect(
			screen.getByRole("list", {
				name: childTaxonomyListName,
			})
		).toBeInTheDocument();
	});

	it("should render a heading for each child taxonomy", () => {
		render(<WoundManagementTaxonomyPage {...props} />);
		const childTaxonomyList = screen.getByRole("list", {
			name: childTaxonomyListName,
		});
		expect(
			within(childTaxonomyList).getAllByRole("heading", { level: 2 })
		).toHaveLength(
			props.data.bnfWoundManagementTaxonomyRoot.taxonomy.childTaxonomies.length
		);
	});
});
