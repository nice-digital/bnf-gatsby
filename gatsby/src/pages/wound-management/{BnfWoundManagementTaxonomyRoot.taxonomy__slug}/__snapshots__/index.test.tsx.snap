// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wound management taxonomy page should match page snapshot 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="https://www.nice.org.uk/"
        >
          NICE
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/"
        >
          BNF
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/wound-management/"
        >
          Wound management products and elasticated garments
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Advanced wound dressings
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "https://www.nice.org.uk/",
        "name": "NICE"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/",
        "name": "BNF"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/wound-management/",
        "name": "Wound management products and elasticated garments"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "Advanced wound dressings"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
    id="content-start"
  >
    <h1
      class="page-header__heading"
    >
      <span>
        Advanced wound dressings
      </span>
    </h1>
  </div>
  <div
    class="grid grid--loose"
    data-testid="body"
  >
    <div
      class="hide-print"
      data-g="12 md:4 lg:3"
    >
       
      <button
        aria-expanded="false"
        aria-label="Expand menu for wound management products pages"
        class="toggleButton"
        type="button"
      >
        <span>
          Wound management products
        </span>
         
        <svg
          aria-hidden="true"
          class="icon"
          fill="none"
          focusable="false"
          height="1em"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M478.152 174c0 10.091-3.525 18.661-10.576 25.712l-186 186c-7.243 7.243-15.909 10.864-26 10.864-10.283 0-18.853-3.621-25.712-10.864l-186-186C36.621 192.853 33 184.283 33 174c0-10.091 3.621-18.757 10.864-26L65 126.576C72.424 119.525 81.09 116 91 116c10.091 0 18.661 3.525 25.712 10.576L255.576 265.44 394.44 126.576c7.051-7.051 15.621-10.576 25.712-10.576 9.909 0 18.576 3.525 26 10.576L467.576 148c7.051 7.424 10.576 16.091 10.576 26z"
            fill="currentColor"
          />
        </svg>
      </button>
      <nav
        aria-label="Wound management products pages"
        class="stacked-nav"
        id="collapsible-menu"
      >
        <p
          class="stacked-nav__root"
        >
          <a
            aria-current="false"
            href="/wound-management/"
          >
            Wound management products
          </a>
        </p>
        <ul
          class="stacked-nav__list"
        >
          <li
            class="stacked-nav__list-item"
          >
            <a
              aria-current="false"
              href="/wound-management/basic-wound-contact-dressings/"
            >
              <span
                class="stacked-nav__content-wrapper"
              >
                <span>
                  Basic wound contact dressings
                </span>
              </span>
            </a>
          </li>
          <li
            class="stacked-nav__list-item"
          >
            <a
              aria-current="false"
              href="/wound-management/advanced-wound-dressings/"
            >
              <span
                class="stacked-nav__content-wrapper"
              >
                <span>
                  Advanced wound dressings
                </span>
              </span>
            </a>
          </li>
        </ul>
      </nav>
    </div>
    <div
      data-g="12 md:8 lg:9"
    >
      <nav
        aria-labelledby="navigate-to-section"
        class="nav"
      >
        <h2
          class="heading"
          id="navigate-to-section"
        >
          Navigate to section
        </h2>
        <ol
          aria-label="Jump links to sections on this page"
          class="linkList"
        >
          <li>
            <a
              href="#capillary-acting-dressings"
            >
              Capillary-acting dressings
            </a>
          </li>
          <li>
            <a
              href="#hydrogel-dressings"
            >
              Hydrogel dressings
            </a>
          </li>
          <li>
            <a
              href="#vapour-permeable-films-and-membranes"
            >
              Vapour-permeable films and membranes
            </a>
          </li>
        </ol>
      </nav>
      <div>
        <p>
           Advanced wound dressings can be used for both acute and chronic wounds. Categories for dressings in this section start with the least absorptive, moisture-donating hydrogel dressings, followed by increasingly more absorptive dressings. 
        </p>
      </div>
      <ul
        aria-label="List of advanced wound dressings"
        class="childTaxonomyList"
      >
        <li>
          <h2
            id="capillary-acting-dressings"
          >
            Capillary-acting dressings
          </h2>
          <ul
            class="nestedTaxonomyList"
          >
            <li>
              <a
                href="/wound-management/advanced-wound-dressings/capillary-acting-dressings/"
              >
                Capillary-acting dressings
              </a>
            </li>
          </ul>
        </li>
        <li>
          <h2
            id="hydrogel-dressings"
          >
            Hydrogel dressings
          </h2>
          <div>
            <p>
               Hydrogel dressings are most commonly supplied as an amorphous, cohesive topical application that can take up the shape of a wound.
            </p>
          </div>
          <ul
            class="nestedTaxonomyList"
          >
            <li>
              <a
                href="/wound-management/advanced-wound-dressings/hydrogel-application-amorphous/"
              >
                Hydrogel application (amorphous)
              </a>
            </li>
            <li>
              <a
                href="/wound-management/advanced-wound-dressings/hydrogel-sheet-dressings/"
              >
                Hydrogel sheet dressings
              </a>
            </li>
            <li>
              <a
                href="/wound-management/advanced-wound-dressings/sodium-hyaluronate-dressings/"
              >
                Sodium hyaluronate dressings
              </a>
            </li>
          </ul>
        </li>
        <li>
          <h2
            id="vapour-permeable-films-and-membranes"
          >
            Vapour-permeable films and membranes
          </h2>
          <div>
            <p>
               Vapour-permeable films and membranes allow the passage of water vapour and oxygen but are impermeable to water and micro-organisms, and are suitable for lightly exuding wounds.
            </p>
          </div>
          <ul
            class="nestedTaxonomyList"
          >
            <li>
              <a
                href="/wound-management/advanced-wound-dressings/non-woven-fabric-dressing-with-viscose-rayon-pad/"
              >
                Non-woven fabric dressing with viscose-rayon pad
              </a>
            </li>
            <li>
              <a
                href="/wound-management/advanced-wound-dressings/vapour-permeable-adhesive-film-dressing-semi-permeable-adhesive-dressing/"
              >
                Vapour-permeable Adhesive Film Dressing (Semi-permeable Adhesive Dressing)
              </a>
            </li>
            <li>
              <a
                href="/wound-management/advanced-wound-dressings/vapour-permeable-transparent-film-dressing-with-adhesive-foam-border/"
              >
                Vapour–permeable transparent film dressing with adhesive foam border
              </a>
            </li>
            <li>
              <a
                href="/wound-management/advanced-wound-dressings/vapour-permeable-transparent-adhesive-film-dressing/"
              >
                Vapour–permeable transparent, adhesive film dressing.
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
